src/test/allocator_tests.cpp

```diff
@@ -11,110 +11,224 @@
-    lpm.UnlockRange(reinterpret_cast<void*>(test_page_size*1200), test_page_size*500-1);
+private:
+    int count;
+    int lockedcount;
+};
 
-    /* Check that everything is released */
-    BOOST_CHECK(lpm.GetLockedPageCount() == 0);
+BOOST_AUTO_TEST_CASE(lockedpool_tests_mock)
+{
+    // Test over three virtual arenas, of which one will succeed being locked
+    std::unique_ptr<LockedPageAllocator> x(new TestLockedPageAllocator(3, 1));
+    LockedPool pool(std::move(x));
+    BOOST_CHECK(pool.stats().total == 0);
+    BOOST_CHECK(pool.stats().locked == 0);
 
-    /* A few and unlocks of size zero (should have no effect) */
-    addr = 0;
-    for(int i=0; i<1000; ++i)
-    {
-        lpm.LockRange(reinterpret_cast<void*>(addr), 0);
-        addr += 1;
-    }
-    BOOST_CHECK(lpm.GetLockedPageCount() == 0);
-    addr = 0;
-    for(int i=0; i<1000; ++i)
+    // Ensure unreasonable requests are refused without allocating anything
+    void *invalid_toosmall = pool.alloc(0);
+    BOOST_CHECK(invalid_toosmall == nullptr);
+    BOOST_CHECK(pool.stats().used == 0);
+    BOOST_CHECK(pool.stats().free == 0);
+    void *invalid_toobig = pool.alloc(LockedPool::ARENA_SIZE+1);
+    BOOST_CHECK(invalid_toobig == nullptr);
+    BOOST_CHECK(pool.stats().used == 0);
+    BOOST_CHECK(pool.stats().free == 0);
+
+    void *a0 = pool.alloc(LockedPool::ARENA_SIZE / 2);
+    BOOST_CHECK(a0);
+    BOOST_CHECK(pool.stats().locked == LockedPool::ARENA_SIZE);
+    void *a1 = pool.alloc(LockedPool::ARENA_SIZE / 2);
+    BOOST_CHECK(a1);
+    void *a2 = pool.alloc(LockedPool::ARENA_SIZE / 2);
+    BOOST_CHECK(a2);
+    void *a3 = pool.alloc(LockedPool::ARENA_SIZE / 2);
+    BOOST_CHECK(a3);
+    void *a4 = pool.alloc(LockedPool::ARENA_SIZE / 2);
+    BOOST_CHECK(a4);
+    void *a5 = pool.alloc(LockedPool::ARENA_SIZE / 2);
+    BOOST_CHECK(a5);
+    // We've passed a count of three arenas, so this allocation should fail
+    void *a6 = pool.alloc(16);
+    BOOST_CHECK(!a6);
+
+    pool.free(a0);
+    pool.free(a2);
+    pool.free(a4);
+    pool.free(a1);
+    pool.free(a3);
+    pool.free(a5);
+    BOOST_CHECK(pool.stats().total == 3*LockedPool::ARENA_SIZE);
+    BOOST_CHECK(pool.stats().locked == LockedPool::ARENA_SIZE);
+    BOOST_CHECK(pool.stats().used == 0);
+}
+
+// These tests used the live LockedPoolManager object, this is also used
+// by other tests so the conditions are somewhat less controllable and thus the
+// tests are somewhat more error-prone.
+BOOST_AUTO_TEST_CASE(lockedpool_tests_live)
+{
+    LockedPoolManager &pool = LockedPoolManager::Instance();
+    LockedPool::Stats initial = pool.stats();
+
+    void *a0 = pool.alloc(16);
+    BOOST_CHECK(a0);
+    // Test reading and writing the allocated memory
+    *((uint32_t*)a0) = 0x1234;
+    BOOST_CHECK(*((uint32_t*)a0) == 0x1234);
+
+    pool.free(a0);
+    try { // Test exception on double-free
+        pool.free(a0);
+        BOOST_CHECK(0);
+    } catch(std::runtime_error &)
```

Comment:
The error you suggested be turned into an assertion (which I agree is a clear programming error, to call `LockedPool::free` on a pointer not returned by `LockedPool::alloc`) does not trigger this. The pointer is within an arena in both calls to `pool.free`, so it is `Arena::free` that checks for the double-free and throws an exception.

Reply:
Ah ok. Then perhaps `Arena::free` should also be aborting in that case. Exceptions aren't a good way to report that error.

Reply:
Aha, I've found what I'm after: https://stackoverflow.com/a/34676852/3317191 points to the C++14 Table 28 Allocator requirements, which imply that an exception during `deallocate` is a violation of the required contact, resulting in undefined behaviour.

Reply:
I've found the [named requirements for Allocator](https://en.cppreference.com/w/cpp/named_req/Allocator):

Expression | Return type | Requirements
-- | -- | --
`a.deallocate(p, n)` | _(not used)_ | Deallocates storage pointed to `p`, which must be a value returned by a previous call to `allocate` that has not been invalidated by an intervening call to `deallocate`. `n` must match the value previously passed to `allocate`. Does not throw exceptions.

This excludes both:
- Calling `deallocate` on `nullptr` (which `LockedPool::free` is documented as supporting, although it doesn't currently work; only `Arena::free(nullptr)` succeeds, but `nullptr` isn't part of any arena).
- Throwing exceptions of any kind (even if a double-free can be safely detected and prevented, as in `Arena::free`).

Note that these requirements technically apply to `secure_allocator`, not `LockedPool` itself. We could leave `LockedPool` as-is and encapsulate these cases (by adding `noexcept` to secure_allocator::deallocate`'s declaration and asserting in it that `p != nullptr`). However `secure_allocator` simply wraps `LockedPool` and adds zeroizing of memory on deallocation, so it might make sense to apply the deallocation-relevant named requirements to `LockedPool` and `Arena`.

Reply:
It's "traditional" to allow `nullptr` as an argument to `free`-like functions and make it a no-op, so I think we should allow that for direct calls to `LockedPool::free` and make it work as documented. Attempting to free a detectably invalid pointer should terminate. There doesn't appear to be anything that would prohibit doing that by `assert(!"message");`, but the test should confirm that it works.

Reply:
Also we should submit this change upstream.